CRM/Core/BAO/RecurringEntity.php

```diff
@@ -596,6 +596,14 @@ static public function copyCreateEntity($entityTable, $fromCriteria, $newParams,
       $object->find(TRUE);
 
       CRM_Core_BAO_RecurringEntity::quickAdd($object->id, $newObject->id, $entityTable);
+      CRM_Core_DAO::copyCustomFields($object->id, $newObject->id, $entityTable);
+      CRM_Core_DAO::copyGeneric('CRM_Core_DAO_EntityTag',
+        array(
+          'entity_id' => $object->id,
+          'entity_table' => $entityTable,
+        ),
+        array('entity_id' => $newObject->id)
+      );
```

Comment:
By adding these two lines here we are saying: "For any entity copy the tags and customfields".  This is probably ok but I'd rather see movement away from recurringentity knowing anything about the actual entities and instead falling back to entity specific bao methods which understand exactly what is needed to copy an entity in one place (we currently duplicate most of the event copy in recurring entity which is messy.
Quite a few entities already have a copy function (eg. CRM_Event_BAO_Event::copy), how about creating a "copy" function on CRM_Core_DAO for any entities that don't need any special handling, trying to use the entity specific BAO copy function first and falling back to the CRM_Core_DAO method otherwise?
Eg.
```
if (method_exists($baoName, 'copy')) {
  $baoName::copy(..);
}
else {
  CRM_Core_DAO::copy(...);
}
```

CRM_Core_DAO::copy does:
```
CRM_Core_DAO::copyGeneric($daoName, $fromCriteria, $newParams);
CRM_Core_DAO::copyCustomFields($object->id, $newObject->id, $entityTable);
CRM_Core_DAO::copyGeneric('CRM_Core_DAO_EntityTag',
```
and anything else that is generic to ALL entities.

Reply:
@mattwire Don't all the BAO classes inherit from `CRM_Core_DAO` anyway?  So that if statement would be redundant?

Reply:
@agileware-fj Yes you are right :-) We don't need the if statement.

Reply:
@mattwire Before we implement this I would like discuss some technical difficulties we're facing. 

As per your comment we need to move the above three functions in CRM_Core_DAO::copy function. But as you've mentioned some of the classes already implements copy function of their own.

If we want to keep the logic in common function we will need to have common parameters in method. So that would require an interface change. Also the copyCreateEntity function does copy the generic entity before executing the custom fields and tags copy methods. So we also need to think on that because the specific BAO classes also does that in their own copy function. 

I guess we need to discuss and finalise the implementation before we change the current implementation. 

I think we can have that path in another PR too as this one fixes the mentioned issue.